### PR TITLE
4412 populate rows

### LINKED ISSUE #4412: Create R&R forms, prepopulated with Open mSupply data
Add a create button to the list view.

This should open a modal, where you can select the program, and the period.

You can only create an R&R form for a given period once the form for the previous period has been submitted. So the earliest period without a form should be automatically selected.

If no R&R forms for that master list exist yet, select the earliest period.

R&R form should only be able to be created if the previous one is submitted. (TODO: confirm this - and what do we do for the first one?)

From the given Master List, for each item, create R&R row:

- Item name (from item)
- strength (from item)
- unit (from item)
- VEN (from item - will require translation from mSupply)
- initial balance (at the start of the period, from item ledger)
- quantity received  (within the period, from item ledger)
- quantity consumed (within the period, from item ledger)
- adjusted quantity consumed - (quantity consumed x 30)/(30 - stock out duration) _(30, or days in month?)_
- losses/adjustments - blank, to be entered
- stock out duration (days) - (could be calculated... # days where stock = 0)
- final balance -  (initial balance) + (quantity received) - (quantity consumed) ± (losses / adjustments)
- AMC
- Maximum quantity - AMC * 2
- expiration date  - blank, to be entered
- requested quantity - Maximum quantity - final balance
- Approved quantity - blank, to be entered
- comment - blank, to be entered

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #4412

# 👩🏻‍💻 What does this PR do?

Generates data for the R&R form rows

## 💌 Any notes for the reviewer?

Notes throughout :)

# 🧪 Testing

- [ ] Run tests
- [ ] Could have some transactions for items, then create R&R - check DB that prepopulated values reflect usage ... this will be easier once there is a UI though!

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: server/repository/src/db_diesel/adjustment.rs
Comment (lache-melvin) [inline server/repository/src/db_diesel/adjustment.rs:10]: This view and `replenishment` are equivalents of the consumption view, for other invoice types

File: server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs
Comment (lache-melvin) [inline server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs:7]: Dropping here to add Prescription invoice type

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:19]: Future future :))

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:81]: I.e. lookback months is just period_days/days_per_month...

Could use item_stats AMC, but that is current AMC for the item, rather than AMC at the time of R&R period...

Perhaps its fine for now, can expand on AMC in future?

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:155]: There is a risk here, if there is a gap between previous R&R form period end date, and this period start date...

Is that worth checking here?

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:33]: This is what we have done for AMC throughout the rest of the app... however, there was mention in the R&R forms of getting actual days per month for the period - should I do that here?

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:147]: Lol maybe some of these helper methods should be in their own file 😅

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:210]: Wanted to reuse stock evolution, but this is a bit ick... you reckon its fine?

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:235]: doing a lot of casting between different number types... is that risky?

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [inline server/service/src/rnr_form/generate_rnr_form_lines.rs:313]: Is this ok? I guess would be a little strange to show currently available earliest expiring stock if you're filling forms in retroactively, but is that fairly unlikely?

**Review by jmbrunskill — COMMENTED** [review]
Great job, so many things to think through!

I feel like I've raised a lot of questions and not sure if I'm quite ready to approve it yet, Leaning towards approving it and coming back to questions but maybe can revisit on monday?

Most of the comments are really suggestions or questions, nothing critical I don't think.

Might be good to get another pair of eyes on it too?

Lets catch up after you've had a chance to look at the comments.

**Review by jmbrunskill — APPROVED** [review]
Approving based on conversations today.
Nothing blocking merging for now...

File: server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs
Comment (lache-melvin) [reply server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs:28]: #4462 created

File: server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs
Comment (lache-melvin) [reply server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs:40]: #4460 created

File: server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs
Comment (lache-melvin) [reply server/repository/src/migrations/v2_02_00/consumption_and_replenishment_views.rs:63]: That's a cool idea! Have linked this comment in #4462 to come back to during a refactor 🚀 

As mentioned in person `Repack` wasn't included as it doesn't impact _item_ quantity, only an individual stock line, which is not included in this view. I'll add a comment about that though!

**Comment by mark-prins** [comment]
> we could just get the consumption from the x previous R&R forms and average them? Ignoring all the complication of existing transactions & calculation methods?

yes - we should do that. If using R&R, then the AMC is all in the R&R forms, that way we cater for R&R only sites as well as sites that use omSupply for stock management

File: server/service/src/rnr_form/insert.rs
Comment (lache-melvin) [reply server/service/src/rnr_form/insert.rs:144]: Yep, but added in next PR :)

File: server/service/src/rnr_form/generate_rnr_form_lines.rs
Comment (lache-melvin) [reply server/service/src/rnr_form/generate_rnr_form_lines.rs:238]: I think will keep this, as it handles if the period is longer than a month...

**Comment by Chris-Petty** [comment]
Random note - I got confused by changes here in the code base for a second because it adds a replenishment view while OG has a replenishment table!

image

Replenishments (OG) being essentially warehouses' instructions to move stock from bulk locations to pick faces, i.e. replenishing the pick faces.

image

**Pull request merged**
